File: lib/p2p/chain/action/signal.ts
import {ChainMessage, MsgType} from '@/libs/p2p/message';
import {BaseAction} from '@/libs/p2p/chain/baseaction';
import {webrtcPeerPool} from '@/libs/transport/webrtc/webrtcpeerpool';

/**
 在chain目录下的采用自定义protocol "/chain"的方式自己实现的功能
 */
export class SignalAction extends BaseAction {
  constructor(msgType: MsgType) {
    super(msgType);
    webrtcPeerPool.registSignalAction(this);
  }

  async signal(connectPeerId: string, data: any, targetPeerId: string): Promise<any> {
    let chainMessage: ChainMessage = signalAction.prepareSend(connectPeerId, data, targetPeerId);
    // TODO: 视频通话的signal加密送过去解密完数据有问题，具体原因还没找到
    //chainMessage.NeedEncrypt = true

    let response: ChainMessage = await signalAction.send(chainMessage);
    if (response) {
      return response.Payload;
    }

    return null;
  }

  async receive(chainMessage: ChainMessage): Promise<ChainMessage> {
    let _chainMessage: ChainMessage = await super.receive(chainMessage);
    if (_chainMessage && signalAction.receivers) {
      signalAction.receivers.forEach((receiver, key) => {
        if (_chainMessage) {
          receiver(_chainMessage.SrcPeerId, _chainMessage.SrcConnectPeerId, _chainMessage.SrcConnectSessionId, _chainMessage.Payload);
        }
      });

      return null;
    }
  }
}

export let signalAction = new SignalAction(MsgType.SIGNAL);

File: lib/p2p/chain/action/signal.dart
import 'package:colla_chat/entity/p2p/chain_message.dart';
import 'package:colla_chat/p2p/chain/baseaction.dart';
import 'package:colla_chat/transport/webrtc/base_peer_connection.dart';

///这是webrtc信号服务器的客户端实现，可以发送和接收信号服务器的数据
class SignalAction extends BaseAction {
  SignalAction(MsgType msgType) : super(msgType);

  ///发送webrtc信号
  Future<bool> signal(WebrtcSignal signal, String targetPeerId,
      {String? targetClientId}) async {
    ChainMessage? chainMessage = await signalAction.prepareSend(signal,
        targetPeerId: targetPeerId, targetClientId: targetClientId);
    // TODO: 视频通话的signal加密送过去解密完数据有问题，具体原因还没找到
    chainMessage.needEncrypt = false;
    chainMessage.needCompress = false;

    return await signalAction.send(chainMessage);
  }

  @override
  Future<void> response(ChainMessage chainMessage) async {}

  @override
  Future<void> transferPayload(ChainMessage chainMessage) async {
    super.transferPayload(chainMessage);
    var json = chainMessage.payload;
    WebrtcSignal signal = WebrtcSignal.fromJson(json);
    chainMessage.payload = signal;
  }
}

final signalAction = SignalAction(MsgType.SIGNAL);
